plugins/automation/app/controllers/discourse_automation/admin_automations_controller.rb

```diff
@@ -50,31 +53,36 @@ def update
           last_updated_by_id: current_user.id,
         )
 
-      if automation.trigger != params[:automation][:trigger]
+      if attributes.key?(:trigger) && automation.trigger != params[:automation][:trigger]
         params[:automation][:fields] = []
         attributes[:enabled] = false
         automation.fields.destroy_all
       end
 
-      if automation.script != params[:automation][:script]
-        attributes[:trigger] = nil
-        params[:automation][:fields] = []
-        attributes[:enabled] = false
-        automation.fields.destroy_all
-        automation.tap { |r| r.assign_attributes(attributes) }.save!(validate: false)
+      if attributes.key?(:script)
+        if automation.script != params[:automation][:script]
+          attributes[:trigger] = nil
+          params[:automation][:fields] = []
+          attributes[:enabled] = false
+          automation.fields.destroy_all
+          automation.tap { |r| r.assign_attributes(attributes) }.save!(validate: false)
+        else
+          Array(params[:automation][:fields])
+            .reject(&:empty?)
+            .each do |field|
+              automation.upsert_field!(
+                field[:name],
+                field[:component],
+                field[:metadata],
+                target: field[:target],
+              )
+            end
+
+          automation.tap { |r| r.assign_attributes(attributes) }.save!
```

Comment:
I believe if we don't need the `validate: false` option to save, we can use `#update!`:

```suggestion
          automation.update!(attributes)
```